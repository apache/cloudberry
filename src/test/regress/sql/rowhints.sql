-- Test Optimizer Row Hints Feature
--
-- Purpose: Test that row hints may be used to edit cardinality estimates

LOAD 'pg_hint_plan';

DROP SCHEMA IF EXISTS rowhints CASCADE;

CREATE SCHEMA rowhints;
SET search_path=rowhints;
SET optimizer_trace_fallback=on;

-- Setup tables
CREATE TABLE my_table(a int, b int);
CREATE INDEX my_awesome_index ON my_table(a);

CREATE TABLE your_table(a int, b int) WITH (appendonly=true);
CREATE INDEX your_awesome_index ON your_table(a);

CREATE TABLE our_table(a int, b int) PARTITION BY RANGE (a) (PARTITION p1 START(0) END(100) INCLUSIVE EVERY(20));
CREATE INDEX our_awesome_index ON our_table(a);

INSERT INTO my_table SELECT i, i FROM generate_series(1, 100)i;
INSERT INTO your_table SELECT i, i FROM generate_series(1, 100)i;
INSERT INTO our_table SELECT i, i FROM generate_series(1, 100)i;
ANALYZE my_table, your_table, our_table;

--------------------------------------------------------------------
-- Test the different row hint types:
--
--     - Absolute
--     - Add
--     - Subtract
--     - Multiply
--------------------------------------------------------------------

-- Baseline no hints
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 #123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 +123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 -123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 *123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout


--------------------------------------------------------------------
--
-- Subqueries
--
--------------------------------------------------------------------

\o results/pg_hint_plan.tmpout
/*+
    Rows(my_table your_table #123)
 */
EXPLAIN SELECT * FROM my_table, (SELECT * FROM your_table) AS q;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout


--------------------------------------------------------------------
--
-- CTE
--
--------------------------------------------------------------------

\o results/pg_hint_plan.tmpout
/*+
    Rows(my_table your_table #123)
 */
EXPLAIN WITH cte AS (SELECT * FROM my_table, (SELECT * FROM your_table) as q) SELECT * FROM cte;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout


--------------------------------------------------------------------
-- Test updating lower join row hint
--------------------------------------------------------------------

-- force join order to isolate lower join row hint
set optimizer_join_order=query;

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 #123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 *123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
