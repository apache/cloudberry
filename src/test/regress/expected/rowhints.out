-- Test Optimizer Row Hints Feature
--
-- Purpose: Test that row hints may be used to edit cardinality estimates
LOAD 'pg_hint_plan';
DROP SCHEMA IF EXISTS rowhints CASCADE;
CREATE SCHEMA rowhints;
SET search_path=rowhints;
SET optimizer_trace_fallback=on;
-- Setup tables
CREATE TABLE my_table(a int, b int);
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE INDEX my_awesome_index ON my_table(a);
CREATE TABLE your_table(a int, b int) WITH (appendonly=true);
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE INDEX your_awesome_index ON your_table(a);
CREATE TABLE our_table(a int, b int) PARTITION BY RANGE (a) (PARTITION p1 START(0) END(100) INCLUSIVE EVERY(20));
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE INDEX our_awesome_index ON our_table(a);
INSERT INTO my_table SELECT i, i FROM generate_series(1, 100)i;
INSERT INTO your_table SELECT i, i FROM generate_series(1, 100)i;
INSERT INTO our_table SELECT i, i FROM generate_series(1, 100)i;
ANALYZE my_table, your_table, our_table;
--------------------------------------------------------------------
-- Test the different row hint types:
--
--     - Absolute
--     - Add
--     - Subtract
--     - Multiply
--------------------------------------------------------------------
-- Baseline no hints
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
                                                 QUERY PLAN                                                  
-------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=20000000000.00..20000017562.79 rows=1000000 width=8)
   ->  Nested Loop  (cost=20000000000.00..20000004229.45 rows=333333 width=8)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..2.67 rows=100 width=4)
               ->  Seq Scan on your_table t2  (cost=0.00..1.33 rows=33 width=4)
         ->  Materialize  (cost=10000000000.00..10000000068.45 rows=3333 width=4)
               ->  Nested Loop  (cost=10000000000.00..10000000051.79 rows=3333 width=4)
                     ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=0.00..7.87 rows=102 width=0)
                           ->  Append  (cost=0.00..6.51 rows=34 width=0)
                                 ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=0.00..1.06 rows=6 width=0)
                                 ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=0.00..1.07 rows=7 width=0)
                                 ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=0.00..1.07 rows=7 width=0)
                                 ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=0.00..1.07 rows=7 width=0)
                                 ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=0.00..1.07 rows=7 width=0)
                                 ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=0.00..1.01 rows=1 width=0)
                     ->  Materialize  (cost=0.00..1.50 rows=33 width=4)
                           ->  Seq Scan on my_table t1  (cost=0.00..1.33 rows=33 width=4)
 Optimizer: Postgres-based planner
(17 rows)

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 #123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=xxx..xxx rows=123 width=xxx)
   ->  Nested Loop  (cost=xxx..xxx rows=41 width=xxx)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
               ->  Seq Scan on your_table t2  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost={inf}..{inf} rows=3333 width=xxx)
               ->  Nested Loop  (cost={inf}..{inf} rows=3333 width=xxx)
                     ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=xxx..xxx rows=102 width=xxx)
                           ->  Append  (cost=xxx..xxx rows=34 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=xxx..xxx rows=6 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=xxx..xxx rows=1 width=xxx)
                     ->  Materialize  (cost=xxx..xxx rows=33 width=xxx)
                           ->  Seq Scan on my_table t1  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 +123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=xxx..xxx rows=1000123 width=xxx)
   ->  Nested Loop  (cost=xxx..xxx rows=333374 width=xxx)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
               ->  Seq Scan on your_table t2  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost={inf}..{inf} rows=3333 width=xxx)
               ->  Nested Loop  (cost={inf}..{inf} rows=3333 width=xxx)
                     ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=xxx..xxx rows=102 width=xxx)
                           ->  Append  (cost=xxx..xxx rows=34 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=xxx..xxx rows=6 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=xxx..xxx rows=1 width=xxx)
                     ->  Materialize  (cost=xxx..xxx rows=33 width=xxx)
                           ->  Seq Scan on my_table t1  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 -123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=xxx..xxx rows=999877 width=xxx)
   ->  Nested Loop  (cost=xxx..xxx rows=333292 width=xxx)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
               ->  Seq Scan on your_table t2  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost={inf}..{inf} rows=3333 width=xxx)
               ->  Nested Loop  (cost={inf}..{inf} rows=3333 width=xxx)
                     ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=xxx..xxx rows=102 width=xxx)
                           ->  Append  (cost=xxx..xxx rows=34 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=xxx..xxx rows=6 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=xxx..xxx rows=1 width=xxx)
                     ->  Materialize  (cost=xxx..xxx rows=33 width=xxx)
                           ->  Seq Scan on my_table t1  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 t3 *123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=xxx..xxx rows=123000000 width=xxx)
   ->  Nested Loop  (cost=xxx..xxx rows=41000000 width=xxx)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
               ->  Seq Scan on your_table t2  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost={inf}..{inf} rows=3333 width=xxx)
               ->  Nested Loop  (cost={inf}..{inf} rows=3333 width=xxx)
                     ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=xxx..xxx rows=102 width=xxx)
                           ->  Append  (cost=xxx..xxx rows=34 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=xxx..xxx rows=6 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=xxx..xxx rows=1 width=xxx)
                     ->  Materialize  (cost=xxx..xxx rows=33 width=xxx)
                           ->  Seq Scan on my_table t1  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

--------------------------------------------------------------------
--
-- Subqueries
--
--------------------------------------------------------------------
\o results/pg_hint_plan.tmpout
/*+
    Rows(my_table your_table #123)
 */
EXPLAIN SELECT * FROM my_table, (SELECT * FROM your_table) AS q;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost={inf}..{inf} rows=123 width=xxx)
   ->  Nested Loop  (cost={inf}..{inf} rows=41 width=xxx)
         ->  Seq Scan on my_table  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost=xxx..xxx rows=100 width=xxx)
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
                     ->  Seq Scan on your_table  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

--------------------------------------------------------------------
--
-- CTE
--
--------------------------------------------------------------------
\o results/pg_hint_plan.tmpout
/*+
    Rows(my_table your_table #123)
 */
EXPLAIN WITH cte AS (SELECT * FROM my_table, (SELECT * FROM your_table) as q) SELECT * FROM cte;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost={inf}..{inf} rows=123 width=xxx)
   ->  Nested Loop  (cost={inf}..{inf} rows=41 width=xxx)
         ->  Seq Scan on my_table  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost=xxx..xxx rows=100 width=xxx)
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
                     ->  Seq Scan on your_table  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

--------------------------------------------------------------------
-- Test updating lower join row hint
--------------------------------------------------------------------
-- force join order to isolate lower join row hint
set optimizer_join_order=query;
\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 #123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=xxx..xxx rows=12300 width=xxx)
   ->  Nested Loop  (cost=xxx..xxx rows=4100 width=xxx)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=102 width=xxx)
               ->  Append  (cost=xxx..xxx rows=34 width=xxx)
                     ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=xxx..xxx rows=6 width=xxx)
                     ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=xxx..xxx rows=7 width=xxx)
                     ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=xxx..xxx rows=7 width=xxx)
                     ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=xxx..xxx rows=7 width=xxx)
                     ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=xxx..xxx rows=7 width=xxx)
                     ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=xxx..xxx rows=1 width=xxx)
         ->  Materialize  (cost={inf}..{inf} rows=41 width=xxx)
               ->  Nested Loop  (cost={inf}..{inf} rows=41 width=xxx)
                     ->  Seq Scan on my_table t1  (cost=xxx..xxx rows=33 width=xxx)
                     ->  Materialize  (cost=xxx..xxx rows=100 width=xxx)
                           ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
                                 ->  Seq Scan on your_table t2  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

\o results/pg_hint_plan.tmpout
/*+
    Rows(t1 t2 *123)
 */
EXPLAIN SELECT t1.a, t2.a FROM my_table AS t1, your_table AS t2, our_table AS t3;
\o
\! sql/maskout.sh results/pg_hint_plan.tmpout
  QUERY PLAN
----------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=xxx..xxx rows=123000000 width=xxx)
   ->  Nested Loop  (cost=xxx..xxx rows=41000000 width=xxx)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=xxx..xxx rows=100 width=xxx)
               ->  Seq Scan on your_table t2  (cost=xxx..xxx rows=33 width=xxx)
         ->  Materialize  (cost={inf}..{inf} rows=3333 width=xxx)
               ->  Nested Loop  (cost={inf}..{inf} rows=3333 width=xxx)
                     ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=xxx..xxx rows=102 width=xxx)
                           ->  Append  (cost=xxx..xxx rows=34 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_1 t3  (cost=xxx..xxx rows=6 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_2 t3_1  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_3 t3_2  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_4 t3_3  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_5 t3_4  (cost=xxx..xxx rows=7 width=xxx)
                                 ->  Seq Scan on our_table_1_prt_p1_6 t3_5  (cost=xxx..xxx rows=1 width=xxx)
                     ->  Materialize  (cost=xxx..xxx rows=33 width=xxx)
                           ->  Seq Scan on my_table t1  (cost=xxx..xxx rows=33 width=xxx)
 Optimizer: Postgres-based planner

